packages/babel-traverse/test/path/isPure.js

```diff
@@ -0,0 +1,110 @@
+import { parse } from "@babel/parser";
+
+import _traverse from "../../lib/index.js";
+const traverse = _traverse.default;
+
+function getPath(code) {
+  const ast = parse(code, {
+    plugins: [
+      ["decorators", { version: "2021-12", decoratorsBeforeExport: true }],
+      ["recordAndTuple", { syntaxType: "hash" }],
+      "decoratorAutoAccessors",
+      ["pipelineOperator", { proposal: "hack", topicToken: "%" }],
+    ],
+  });
+  let path;
+  traverse(ast, {
+    Program: function (_path) {
+      path = _path;
+      _path.stop();
+    },
+  });
+  return path;
+}
+
+describe("isPure() returns true", () => {
+  it.each([
+    "class C { [0]() {} }",
+    "class C extends class {} {}",
+    "class C { static accessor x = 1; accessor y = f() }",
+    "class C { #x = f(); static #y }",
+    "class C { static target = new.target }",
+    "class X { get foo() { return 1 } set foo(v) {} }",
+    "class C { static #p = #p in C }",
+  ])(`NodePath(%p).get("body.0").isPure() should be true`, input => {
+    const path = getPath(input).get("body.0");
+    expect(path.node).toBeTruthy();
+    expect(path.isPure()).toBe(true);
+  });
+
+  it.each([
+    "({ x: 1, foo() { return 1 } })",
+    "String.raw`foo`",
+    `"a" + "b"`,
+    `[function () {}]`,
+    `#{ 0: 0, 1n: 1, two: "two"}`,
+    `#[0, 1n, "2", \`3\`]`,
+    `[,]`,
+    `-1 || void 0`,
+    `null ?? (true && false)`,
+    `this`,
+  ])(`NodePath(%p).get("body.0.expression").isPure() should be true`, input => {
+    const path = getPath(input).get("body.0.expression");
+    expect(path.node).toBeTruthy();
+    expect(path.isPure()).toBe(true);
+  });
+
+  it.each(["let a = 1; `${a}`", `let a = 1; a |> % + %`])(
```

Comment:
Uh, the first one is definitely not pure:
```js
let a = { toString() { console.log("Hi :)") } };
`${a}`;
```
however, it was already behaving like this before this PR so if we want to change it I'd prefer to do it in another one.

Reply:
Good point, we can introduce the `pureToString` compiler assumption, just like we have `pureGetter`.